### PR TITLE
PHPLIB-712: Allow hint for unacknowledged writes using OP_MSG when supported by the server

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPLIB-712

POC for https://github.com/mongodb/specifications/pull/1057 ([DRIVERS-1340](https://jira.mongodb.org/browse/DRIVERS-1340))

### CODE REVIEW

File: tests/UnifiedSpecTests/ExpectedError.php [hunk 1]
```diff
@@ -176,6 +172,21 @@ public function assert(?Throwable $e = null): void
         }
     }
 
+    private function assertIsClientError(Throwable $e): void
+    {
+        /* Note: BulkWriteException may proxy a previous exception. Unwrap it
+         * to check the original error. */
+        if ($e instanceof BulkWriteException && $e->getPrevious() !== null) {
+            $e = $e->getPrevious();
+        }
```
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/ExpectedError.php:181]: Explicit UnsupportedExceptions are now implemented and tested for Update, Delete, and FindAndModify operations.

File: src/Operation/Update.php [hunk 2]
```diff
@@ -55,7 +55,10 @@ class Update implements Executable, Explainable
     private static $wireVersionForDocumentLevelValidation = 4;
 
     /** @var integer */
-    private static $wireVersionForHintServerSideError = 5;
+    private static $wireVersionForHint = 8;
```
[the remark this answers is not in the document]
Comment (jmikola) [reply src/Operation/Update.php:59]: Implemented `is_write_concern_acknowledged`, which is now used in the Update, Delete, and FindAndModify operations. PHPLIB-709 has been fixed in the process.

**Comment by jmikola** [comment]
Note: https://github.com/mongodb/specifications/pull/1057#discussion_r697775612 explains the failing CI build. I think this will require removing `outcome` assertions from the tests that execute `w:0` writes, but I'm waiting on some feedback in the spec PR before making that change.

File: tests/Operation/DeleteFunctionalTest.php [hunk 3]
```diff
@@ -63,6 +64,46 @@ public function testDeleteMany(): void
         $this->assertSameDocuments($expected, $this->collection->find());
     }
 
+    public function testHintOptionUnsupportedClientSideError(): void
```
Comment (alcaeus) [inline tests/Operation/DeleteFunctionalTest.php:67]: Are these scenarios not covered by the automated spec tests, or did you want to ensure that the correct exception thrown?

**Review by alcaeus — COMMENTED** [review]
LGTM pending spec test merge.

File: tests/UnifiedSpecTests/ExpectedError.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply tests/UnifiedSpecTests/ExpectedError.php:181]: This addresses [PHPLIB-711](https://jira.mongodb.org/browse/PHPLIB-711).

File: tests/Operation/DeleteFunctionalTest.php (hunk 3 above)
Comment (jmikola) [reply tests/Operation/DeleteFunctionalTest.php:67]: Had I not implemented [PHPLIB-711](https://jira.mongodb.org/browse/PHPLIB-711), we could rely on the spec tests to differentiate between PHPLIB's UnsupportedException and the BulkWriteException thrown from PHPC encountering libmongoc's client-side error. However, I think unwrapping BulkWriteExceptions is necessary to properly differentiate cilent-side and server-side errors.

With PHPLIB-711 implemented, the unwrapped BulkWriteException (in this particular case) and UnsupportedException are both considered client-side errors so the spec tests really cannot differentiate. That's where these tests come in, to explicitly assert that PHPLIB raises its own internal exception.

**Pull request merged**